src/librustc_mir/transform/type_check.rs

```diff
@@ -1070,27 +1076,36 @@ impl<'a, 'gcx, 'tcx> TypeChecker<'a, 'gcx, 'tcx> {
                                         rv,
                                         "accessed field #{} but variant only has {}",
                                         i,
-                                        field_count);
+                                        field_count
+                                    );
                                     continue;
-                                },
+                                }
                             };
                             let op_ty = match op {
-                                Operand::Consume(lv) => lv.ty(mir, tcx).to_ty(tcx),
+                                Operand::Consume(lv) => {
+                                    self.normalize(&lv.ty(mir, tcx), location).to_ty(tcx)
```

Comment:
Very interesting. It's not obvious to me why this is necessary, though it doesn't seem *bad* per se.

Reply:
OK, I see, it's needed only in the "shim MIR" that we construct -- e.g., the MIR that we build for the constructor of a type like `struct Foo(X)`. In that code (`librustc_mir/shim.rs`), we are not presently normalizing associated types as I would expect.

Reply:
I think we'll leave it like this but open a FIXME.